File: study0/solutions/week5/fair_candy_swap.py
class Solution:
    def fairCandySwap(self, aliceSizes: List[int], bobSizes: List[int]) -> List[int]:
        alice_sum = 0
        bob_sum = 0
        bob_candies_idx = set()
        for candies in aliceSizes:
            alice_sum += candies

        for candies in bobSizes:
            bob_sum += candies
            bob_candies_idx.add(candies)
        
        diff = int((alice_sum - bob_sum)/2)

        for candies in aliceSizes:
            if (candies - diff) in bob_candies_idx:
                return [candies, candies - diff]

File: study1/week5/fair-candy-swap.java
//https://leetcode.com/problems/fair-candy-swap/description/

class Solution {
    public int[] fairCandySwap(int[] aliceSizes, int[] bobSizes) {
        //time complexity is O(n+m),where n is the size of aliceSizes and m is the size of bobSizes
        //space complexity is O(1)
        
        int aSum = sum(aliceSizes);
        int bSum = sum(bobSizes);
        int diff = (aSum - bSum) / 2;

        int[] result = new int[2];

        HashSet<Integer> bobSizesSet = new HashSet<Integer>();
        for (int bNum : bobSizes) {
            bobSizesSet.add(bNum);
        }
        
        for (int aNum : aliceSizes) {
            if (bobSizesSet.contains(aNum - diff)) {
                result[0] = aNum;
                result[1] = aNum - diff;
                break;
            }
        }
        return result;
    }

    private int sum(int[] boxes) {
        int sum = 0;
        for (int num : boxes) {
            sum += num;
        }
        return sum;
    }
}
